library/core/src/alloc/layout.rs

```diff
@@ -405,7 +418,9 @@ impl Layout {
     #[stable(feature = "alloc_layout_manipulation", since = "1.44.0")]
     #[inline]
     pub fn array<T>(n: usize) -> Result<Self, LayoutError> {
-        let array_size = mem::size_of::<T>().checked_mul(n).ok_or(LayoutError)?;
+        // Size manipulation is done in isize space to avoid overflowing isize.
+        let n = isize::try_from(n).map_err(|_| LayoutError)?;
```

Comment:
ditto here: 
```rust
Layout::array::<()>(usize::MAX).unwrap() = Layout {
    size: 0,
    align: 1 (1 << 0),
}
```
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=74580b2cdf9e4c3eac2a23b763f1b245

Reply:
This one can add a const collapsible `if size_of::<T>() == 0` to handle this.